contracts/specs/dai.spec

```diff
@@ -0,0 +1,364 @@
+}
+
+// Verify that allowance behaves correctly on approve
+rule approve(address spender, uint256 value) {
+    env e;
+
+    approve(e, spender, value);
+
+    assert(allowance(e.msg.sender, spender) == value, "Approve did not set the allowance as expected");
+}
+
+// Verify revert rules on approve
+rule approve_revert(address spender, uint256 value) {
+    env e;
+
+    approve@withrevert(e, spender, value);
+
+    bool revert1 = e.msg.value > 0;
+
+    assert(revert1 => lastReverted, "Sending ETH did not revert");
+    assert(lastReverted => revert1, "Revert rules are not covering all the cases");
+}
+
+// Verify that allowance behaves correctly on increaseAllowance
+rule increaseAllowance(address spender, uint256 value) {
+    env e;
+
+    uint256 spenderAllowance = allowance(e.msg.sender, spender);
+
+    increaseAllowance(e, spender, value);
+
+    assert(allowance(e.msg.sender, spender) == spenderAllowance + value, "increaseAllowance did not increase the allowance as expected");
+}
+
+// Verify revert rules on increaseAllowance
+rule increaseAllowance_revert(address spender, uint256 value) {
+    env e;
+
+    uint256 spenderAllowance = allowance(e.msg.sender, spender);
+
+    increaseAllowance@withrevert(e, spender, value);
+
+    bool revert1 = spenderAllowance + value > max_uint;
+    bool revert2 = e.msg.value > 0;
+
+    assert(revert1 => lastReverted, "Overflow did not revert");
+    assert(revert2 => lastReverted, "Sending ETH did not revert");
+    assert(lastReverted => revert1 || revert2, "Revert rules are not covering all the cases");
+}
+
+// Verify that allowance behaves correctly on decreaseAllowance
+rule decreaseAllowance(address spender, uint256 value) {
+    env e;
+
+    uint256 spenderAllowance = allowance(e.msg.sender, spender);
+
+    decreaseAllowance(e, spender, value);
+
+    assert(allowance(e.msg.sender, spender) == spenderAllowance - value, "decreaseAllowance did not decrease the allowance as expected");
+}
+
+// Verify revert rules on decreaseAllowance
+rule decreaseAllowance_revert(address spender, uint256 value) {
+    env e;
+
+    uint256 spenderAllowance = allowance(e.msg.sender, spender);
+
+    decreaseAllowance@withrevert(e, spender, value);
+
+    bool revert1 = spenderAllowance - value < 0;
+    bool revert2 = e.msg.value > 0;
+
+    assert(revert1 => lastReverted, "Underflow did not revert");
+    assert(revert2 => lastReverted, "Sending ETH did not revert");
+    assert(lastReverted => revert1 || revert2, "Revert rules are not covering all the cases");
+}
+
+// Verify that supply and balance behave correctly on mint
+rule mint(address to, uint256 value) {
+    env e;
+
+    // Save the totalSupply and sender balance before minting
+    uint256 supply = totalSupply();
+    uint256 toBalance = balanceOf(to);
+    uint256 ward = wards(e.msg.sender);
+
+    require(supply >= toBalance);
```

Comment:
I don't think this decreases assurance since other rules imply it will always hold, but--is it actually needed?

Reply:
Updated using the `requireInvariant` ft. to apply the `balanceSum_equals_totalSupply` invariant.

Reply:
Yeah without it, it failed. If I understand it correctly it is because the balance can be initialized with any value that could be higher than the supply, meaning that the safe math there wouldn't protect from an overflow in the balance violating one of the rules.
As we know the flow of the contract doesn't allow it and now with the invariant we are having the same assumption than with the require but with something that is actually proved is true.